doc/move_group_interface/src/move_group_interface_tutorial.cpp

```diff
@@ -32,169 +32,481 @@
  *  POSSIBILITY OF SUCH DAMAGE.
  *********************************************************************/
 
-/* Author: Sachin Chitta, Dave Coleman, Mike Lautman */
+/* Author: Sachin Chitta, Dave Coleman, Mike Lautman, Simon Goldstein */
 
+// In C++ to use the move_group_interface it and several other files must be included.
+// Both the move_group and planning_scene are accessed through their appropriate interface.
 #include <moveit/move_group_interface/move_group_interface.h>
 #include <moveit/planning_scene_interface/planning_scene_interface.h>
-
+// We also include visual_tools and several other useful message types.
+#include <moveit_visual_tools/moveit_visual_tools.h>
 #include <moveit_msgs/DisplayRobotState.h>
 #include <moveit_msgs/DisplayTrajectory.h>
-
 #include <moveit_msgs/AttachedCollisionObject.h>
 #include <moveit_msgs/CollisionObject.h>
 
-#include <moveit_visual_tools/moveit_visual_tools.h>
+// The rviz_vizual_tools namespace is shortened for code readability.
+namespace rvt = rviz_visual_tools;
 
-int main(int argc, char** argv)
+class MoveGroupInterfaceTutorial
 {
-  ros::init(argc, argv, "move_group_interface_tutorial");
-  ros::NodeHandle node_handle;
-  ros::AsyncSpinner spinner(1);
-  spinner.start();
-
-  // BEGIN_TUTORIAL
+public:
+
+  /**
+  * Basic constucter
+  **/
+  MoveGroupInterfaceTutorial()
+  {
+  }
+  /**
+  * This initializes visual_tools_ and sets the text_pose_ to above the robot
+  */
+  void setupVisualization()
+  {
+    // BEGIN_SUB_TUTORIAL viz_setup
+    //
+    // Visualization
+    // ^^^^^^^^^^^^^
+    //
+    // Within MoveItVisualTools, remote control is an introspection tool that allows users to
+    // step through a high level script via buttons and keyboard shortcuts in RViz.
+    // A variety of commands can be used to display anything from text, points or even paths.
+    //
+    visual_tools_.loadRemoteControl();
+    // END_SUB_TUTORIAL
+
+    // Set the location where text is diplayed to above the robot
+    text_pose_.translation().z() = 1.25;
+  }
+  /**
+  * Publishes the given message above the robot
+  **/
+  void publishToRViz(std::string msg)
+  {
+    visual_tools_.deleteAllMarkers();
+    // To publish a text message inside Rviz
+    visual_tools_.publishText(text_pose_, msg, rvt::WHITE, rvt::XLARGE);
+    // Batch publishing is used to reduce the number of messages being sent to RViz for large visualizations
+    visual_tools_.trigger();
+    // This prompts the user before continuing.
+    visual_tools_.prompt("Press 'next' in the window to continue");
+  }
+
+  /**
+  * Publishes the given message above the robot, and displays a path
+  **/
+  void publishToRViz(std::string msg, const moveit::planning_interface::MoveGroupInterface::Plan& plan)
```

Comment:
No need for copy-pasta here. Call the other `publishToRviz` method or use a pointer for plan and set the default to `nullptr` so that you can optionally pass it in. Then in the code you can check 
```
if (plan) 
{
    visual_tools_.publishTrajectoryLine(plan.trajectory_, joint_model_group_);
}
```

Reply:
see comment about blocking